urldac7fd8c97_vitalymaryin_unsafe_utils_tgz-pJ8-org_vit_unsafeutils_TcpTestJ8/src/org/vit/unsafeutils/tcp/TcpStreamConsumer.java

```diff
@@ -0,0 +1,227 @@
+    private boolean connected = false;
+
+    @Override
+    public void stop()  {
+        started = false;
+        if (socket != null && socket.isOpen())
+            try {
+                socket.close();
+            } catch (IOException e) {
+                logger.error(e.getMessage(), e);
+            }
+    }
+
+    @Override
+    public void start(){
+        makeConnection();
+    }
+
+    @Override
+    public boolean shouldStart() {
+        return true;
+    }
+
+    private void makeConnection(){
+        try {
+            hostAddress = InetAddress.getByName(host);
+            selector = SelectorProvider.provider().openSelector();
+            while (socket == null
+                    || ((!socket.isOpen() || !socket.isConnected()
+                    || !socket.socket().isConnected())
+                    && !socket.isConnectionPending())
+                    ){
+                try {
+                    if(socket == null)
+                        socket = openSocket();
+                    else if (!socket.isOpen() || !socket.isConnected()){
+                        socket.close();
+                        socket = openSocket();
+                    }
+                    logger.info("connecting to:" + hostAddress.getHostName() + ":" + port);
+                    boolean success = socket.connect(new InetSocketAddress(hostAddress, port));
+                    if (!success){
+                        while (!success){
+                            success = socket.finishConnect();
+                        }
+                    }
+                    logger.info("connected to:" + hostAddress.getHostName() + ":" + port);
+                    socket.register(selector, selectionOpts);
+                    pendingData.put(socket, new ConcurrentLinkedQueue<byte[]>());
+                    connected = true;
+                }catch (ConnectException e) {
+                    logger.error(e.getMessage(), e);
+                } catch (ClosedChannelException e) {
+                    logger.error(e.getMessage(), e);
+                } catch (ConnectionPendingException e) {
+                    logger.error(e.getMessage(), e);
+                    break;
+                } catch (IOException e) {
+                    logger.error(e.getMessage(), e);
+                }
+                try {
+                    Thread.sleep(2000);
+                } catch (InterruptedException e) {
+                    logger.error(e.getMessage(), e);
+                }
+            }
+            new Thread(this, "TcpStreamConsumer").start();
+        } catch (UnknownHostException e) {
+            logger.error(e.getMessage(), e);
+        } catch (IOException e) {
+            logger.error(e.getMessage(), e);
+        }
+    }
+
+    protected SocketChannel openSocket() throws IOException {
+        if (debugEnabled)logger.debug("Opening socket");
+        SocketChannel socketChannel = SocketChannel.open();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *socketChannel*. The resource is referenced at lines 121, 122, 123, 124. The resource is returned at line 129. There are other execution paths that don&#39;t close the resource or return it, for example, when AbstractSelectableChannel.configureBlocking throws an exception. To prevent this resource leak, close *socketChannel* along these other paths before you exit this method.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).